finagle-base-http/src/main/scala/com/twitter/finagle/http/headers/Header.scala

```diff
@@ -15,6 +16,20 @@ private[http] object Header {
 
   final class Root private[Header] (name: String, value: String) extends Header(name, value) {
 
+    def iterator: Iterator[HeaderMap.NameValue] =
+      if (next == null) Iterator.single(this)
+      else {
+        var cur: Header = this
```

Comment:
If you move `cur` into the definition of `AbstractIterator` you can avoid lifting it to the the heap via `scala.runtime.ObjectRef`.